File: backend/app/schemas/session_amd.py
"""DTOs mirroring analytics/session_amd.py's dataclasses. Served only by
GET /api/v1/session-amd/{symbol} -- informational, does not feed the
strategy engine, matches the volume-intelligence/volume-profile DTO
pattern.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

SweepDirectionLiteral = Literal["swept_high", "swept_low"]
DistributionDirectionLiteral = Literal["up", "down"]
DistributionStatusLiteral = Literal["Confirmed", "Developing", "Failed"]
CurrentPhaseLiteral = Literal[
    "Accumulating",
    "Range Established -- Awaiting Move",
    "Testing Range",
    "Distribution",
    "Breakout (not manipulation)",
    "No Clear Setup",
]


class AccumulationRangeDTO(BaseModel):
    high: float
    low: float
    range: float
    start_time: datetime
    end_time: datetime
    is_complete: bool


class ManipulationSweepDTO(BaseModel):
    direction: SweepDirectionLiteral
    extreme_price: float
    breakout_time: datetime
    reversal_time: datetime
    candles_to_reverse: int
    expected_distribution_direction: DistributionDirectionLiteral


class DistributionPhaseDTO(BaseModel):
    direction: DistributionDirectionLiteral
    started_at: datetime
    net_move_points: float
    net_move_pct: float
    dominant_side_confirms: bool | None
    status: DistributionStatusLiteral


class SessionAmdDTO(BaseModel):
    symbol: str
    as_of: datetime | None
    accumulation: AccumulationRangeDTO | None
    sweeps: list[ManipulationSweepDTO]
    latest_sweep: ManipulationSweepDTO | None
    distribution: DistributionPhaseDTO | None
    current_phase: CurrentPhaseLiteral
    narrative: str

File: frontend/src/types/sessionAmd.ts
// Mirrors backend/app/schemas/session_amd.py exactly.

export type SweepDirection = "swept_high" | "swept_low";
export type DistributionDirection = "up" | "down";
export type DistributionStatus = "Confirmed" | "Developing" | "Failed";
export type CurrentPhase =
  | "Accumulating"
  | "Range Established -- Awaiting Move"
  | "Testing Range"
  | "Distribution"
  | "Breakout (not manipulation)"
  | "No Clear Setup";

export interface AccumulationRangeDTO {
  high: number;
  low: number;
  range: number;
  start_time: string;
  end_time: string;
  is_complete: boolean;
}

export interface ManipulationSweepDTO {
  direction: SweepDirection;
  extreme_price: number;
  breakout_time: string;
  reversal_time: string;
  candles_to_reverse: number;
  expected_distribution_direction: DistributionDirection;
}

export interface DistributionPhaseDTO {
  direction: DistributionDirection;
  started_at: string;
  net_move_points: number;
  net_move_pct: number;
  dominant_side_confirms: boolean | null;
  status: DistributionStatus;
}

export interface SessionAmdDTO {
  symbol: string;
  as_of: string | null;
  accumulation: AccumulationRangeDTO | null;
  sweeps: ManipulationSweepDTO[];
  latest_sweep: ManipulationSweepDTO | null;
  distribution: DistributionPhaseDTO | null;
  current_phase: CurrentPhase;
  narrative: string;
}
